---
title: "Gene expression analysis"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document:
      df_print: paged
      code_folding: hide
      toc: true
      toc_float: true
      number_sections: true
---

<!-- Set with output width -->
<style type="text/css">
div.main-container {
    max-width: 2000px;
}
.col-md-3 {
    width: 15%;
}
h1 {font-size: 2em}
h2 {font-size: 1.5em}
h3 {font-size: 1.17em}
h4 {font-size: 1.12em}
h5 {font-size: 0.83em}
h6 {font-size: 0.75em}
</style>

```{r setup, include=FALSE}
	# options
	knitr::opts_chunk$set(echo = TRUE)
	options(stringsAsFactors = FALSE)

	# libraries
	library(tidyverse)
	library(cowplot)
	library(fs)
	library(ggrepel)
	library(flextable)
	library(jsonlite)
	library(ggbeeswarm)
	library(GGally)
	library(ggh4x)
	library(ggforce)
	library(ggtext)
	library(officer)
	library(qtl2)
	library(ggupset)
	library(DBI)
	library(dbplyr)

```

# Plotting function

```{r echo=TRUE, fig.width = 6, fig.height = 4, fig.align = 'center', cache = TRUE, eval = TRUE}
	# plotting function
	plt_avg_expr_per_fou_by_gn = function(data) {
		# calculate number of GN datasets per gene to be plotted
		n_gn_per_gene = data %>% distinct(GN, gene_name) %>% count(gene_name)

		# make individual panels
		p_list = data %>%
			mutate(panel_id = gene_name) %>%
			nest(data = !panel_id) %>%
			mutate(p = map(data, function(.data) {
				# center y-axis for each dataset, because expression values are dataset dependent
				# comparing different studies
				if (1) {
					.data = .data %>%
						group_by(GN) %>%
						mutate(expr_val = scale(expr_val, center = TRUE, scale = FALSE)[,1]) %>%
						ungroup
				}

				# only want "B" and "D" genotypes
				.data = .data %>% 
					filter(gt %in% c('0/0', '1/1')) %>%
					mutate(gt = recode(gt, `0/0` = 'B', `1/1` = 'D')) 
				
				# dataset ordering by difference b/w B and D
				dset_lvls = .data %>% 
					group_by(GN, gt) %>%
					summarise(expr_val = mean(expr_val), .groups = 'drop') %>%
					pivot_wider(id_cols = GN, names_from = gt, values_from = expr_val) %>%
					mutate(diff = B - D) %>%
					arrange(desc(diff)) %>% pull(GN) %>% as.character
				
				# add an "all" group
				.data = bind_rows(.data, .data %>% mutate(GN = 'all'))

				# add level for the "all" group
				dset_lvls = c(dset_lvls, 'all')
				
				# form plot
				p = .data %>%
					mutate(GN = fct_relevel(GN, dset_lvls)) %>%
					ggplot(aes(GN, expr_val, color = gt)) +
					geom_boxplot(
								 # outlier.shape = NA, 
								 fill = NA,
								 width = 0.5,
								 position = position_dodge(width = 0.9)) + 
					scale_color_brewer(palette = 'Set1') + 
					scale_x_discrete(guide = guide_axis(check.overlap = TRUE)) +
					scale_y_continuous(guide = guide_axis(check.overlap = TRUE)) +
					facet_grid(cols = vars(gene_name), scales = 'free_x', space = 'free_x') + 
					theme_half_open() +
					theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
						  axis.title.x = element_blank(),
						  legend.position = 'none',
						  plot.title = element_text(size = 10, hjust = 0.5)) +
					labs(y = 'Gene expression')
				p
			}))
		p_list
		# p_list = p_list$p %>% set_names(p_list$panel_id)
		# p = plot_grid(
		# 	plot_grid(p_list$Msh3, p_list$Xrcc4, nrow = 1),
		# 	p_list$Ssbp2, p_list$Atg10, ncol = 1
		# )
		# p_list = p_list %>% left_join(n_gn_per_gene, by = c('panel_id' = 'gene_name'))
		# p_list = p_list %>% mutate(n = n/max(n) + 0.3) 
		# p = plot_grid(plotlist = p_list$p, nrow = 1, rel_widths = p_list$n)
		# p
	}
	
	plot_signif_by_probe = function(all_expr_vals, probe_info, probe_max, gene, n_panel = 2) {
		to_plt = all_expr_vals %>%
			filter(gene_name == gene) %>%
			# number probes
			left_join(probe_info %>% select(probe, probe_pos), by = 'probe') %>%
			group_by(gene_name, GN) %>%
			# order by probe position
			arrange(probe_pos, .by_group = TRUE) %>%
			mutate(probe_num = probe %>% as.character %>% fct_inorder %>% as.integer %>% as.factor) %>%
			ungroup

		# add "is_eqtl" label
		to_plt = to_plt %>%
			left_join(probe_max %>% select(GN, gene_name, probe, is_eqtl), by = c('GN', 'gene_name', 'probe')) %>%
			mutate(is_eqtl = replace_na(is_eqtl, FALSE))

		# group into panels
		panel_ids = to_plt %>%
			distinct(GN) %>%
			mutate(panel_id = cut_number(GN %>% as.factor %>% as.integer, n = n_panel, label = FALSE))

		# make plot list
		p_list = to_plt %>%
			left_join(panel_ids, by = 'GN') %>%
			nest(data = !panel_id) %>%
			mutate(p = map(data, function(.x) {
				.x %>%
				ggplot(aes(probe_num, expr_val, color = is_eqtl)) +
				geom_boxplot(outlier.shape = '.') + 
				facet_grid(~GN, scales = 'free_x', space = 'free') +
				scale_x_discrete(guide = guide_axis(check.overlap = TRUE, n.dodge = 1)) +
				# scale_color_brewer(palette = 'Set1') + 
				scale_color_manual(values = c(`FALSE` = '#377eb8', `TRUE` = '#4daf4a')) + 
				theme_half_open() +
				theme(legend.position = 'none',
					  axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 8),
					  # axis.text.x = element_text(size = 8),
					  strip.text = element_text(angle = 90),
					  panel.spacing = unit(2, 'pt'),
					  text = element_text(size = 8)
				)
			}))
		p = plot_grid(plotlist = p_list$p, ncol = 1)
		p
	}
	redo = TRUE
```

# QTL region

```{r echo=TRUE, fig.width = 6, fig.height = 4, fig.align = 'center', cache = TRUE, eval = TRUE}
	#ci_chr = '13'; ci_lo = 83.78112; ci_hi = 93.41913; ci_mid = 90.4
ci_chr = '13'
qtl_ci = read_csv('../outs/qtl_ci_tetra.csv')
ci_lo = as.double(qtl_ci['ci_lo']); 
ci_hi = as.double(qtl_ci['ci_hi']);
ci_mid = as.double(qtl_ci['pos']);
```

# Other configs

```{r echo=TRUE, fig.width = 6, fig.height = 4, fig.align = 'center', cache = TRUE, eval = TRUE}
	# how many top genes to take
	lod_thresh = 3
	top_x_genes = 10
	other_lvl = str_c('>', top_x_genes)
	
	# define DNA repair genes
	dna_repair_genes = c('Msh3', 'Xrcc4', 'Ssbp2', 'Atg10')	
```

# Load data

#. 54 representative datasets
#. filtered datasets that have >30 strains
#. common protein coding genes

```{r echo=TRUE, fig.width = 6, fig.height = 4, fig.align = 'center', cache = TRUE, eval = TRUE}
	# cached data from eqtl analysis	
	eqtl_data = readRDS('../data/analysis_cache/eqtl_data.rds')
	
	# gene table
	gn_table = readRDS('../data/analysis_cache/final_gn_table.rds') 

	# extract objects
	# best_trace      = eqtl_data$best_trace # want to recalculate best_trace (probes w/ * w/o snps)
	# best_point      = eqtl_data$best_point # want to recalculate best_point (probes w/ * w/o snps)
	sel_dsets         = eqtl_data$sel_dsets
	common_prot_genes = eqtl_data$common_prot_genes
	qtl_dsets         = eqtl_data$qtl_dsets
	eqtl_dsets_genes  = eqtl_data$eqtl_dsets_genes
	gene_ord          = eqtl_data$gene_ord
	probe_max         = eqtl_data$probe_max
	gene_pal          = eqtl_data$gene_pal
	signal_signal_cor = eqtl_data$signal_signal_cor

	# % expanded QTL mapping results
	pheno_vals = read.csv('../outs/pheno_vals_tetra.csv')
	qtl_res = read.csv('../outs/final_qtl_res_tetra.csv')


	# gene expression data
	all_trace     = readRDS('../data/gene_expr/qtl_agg/gene_expr_db.rds')
	all_expr_vals = all_trace$expr_vals
	probe_info    = all_trace$probes

	# filtered for representative datasets and common genes
	# 54 representative datasets already accounted for
	# only use 
	all_expr_vals = all_expr_vals %>%
		filter(gene_name %in% common_prot_genes,
			   GN %in% sel_dsets) %>%
		# filter(GN %in% sel_dsets) %>% # redundant
		semi_join(qtl_dsets, by = 'GN')
	
	# probes with no snps version
	all_expr_vals_w_snp = all_expr_vals # make a backup
	all_expr_vals = all_expr_vals %>%
		semi_join(probe_info %>% filter(n_var_per_probe == 0), by = 'probe')
	
	# best point within QTL sized region around gene
	get_best_point = function(data) {
		ci_hwind = (ci_hi - ci_lo)/2
		data %>%
			left_join(all_trace$genes, by = 'gene_name') %>%
			left_join(all_trace$markers, by = 'marker') %>%
			arrange(desc(LOD)) %>%
			mutate(across(c(gene_pos, gene_end), ~.x/1e6)) %>%
			filter(mark_pos >= (gene_end+gene_pos)/2 - ci_hwind, mark_pos <= (gene_end+gene_pos)/2 + ci_hwind) %>%
			distinct(GN, gene_name, mark_chr, .keep_all = TRUE) %>%
			mutate(across(where(is.factor), as.character))
	}
	best_point = list(
		all_probes = all_trace$signal %>% get_best_point,
		no_snps = all_trace$signal %>%
			semi_join(probe_info %>% filter(n_var_per_probe == 0), by = 'probe') %>%
			get_best_point
	)
```

# Aggregate gene expression data

3 ways of calculating a single expression value per GN/gene/strain

#. Average expression per gene
#. Probe with highest average expression
#. Probe with best QTL signal

```{r echo=TRUE, fig.width = 6, fig.height = 4, fig.align = 'center', cache = TRUE, eval = TRUE}
	# three versions of GN,gene,strain,expr_val dataframes
	unq_expr_vals = list(
		avg_expr_per_gene = all_expr_vals %>%
			semi_join(probe_info %>% distinct(probe, n_var_per_probe) %>% filter(n_var_per_probe == 0), by = 'probe') %>%
			group_by(GN, gene_name, strain) %>%
			summarise(expr_val = mean(expr_val), .groups = 'drop'),
		top_expr_probe = {
			max_expr_probes = all_expr_vals %>%
				semi_join(probe_info %>% distinct(probe, n_var_per_probe) %>% filter(n_var_per_probe == 0), by = 'probe') %>%
				group_by(GN, gene_name, probe) %>%
				summarise(avg_expr = mean(expr_val), .groups = 'drop') %>%
				group_by(GN, gene_name) %>%
				slice_max(n = 1, order_by = avg_expr, with_ties = FALSE) %>% 
				ungroup 
			all_expr_vals %>%
				semi_join(probe_info %>% distinct(probe, n_var_per_probe) %>% filter(n_var_per_probe == 0), by = 'probe') %>%
				semi_join(max_expr_probes, by = c('GN', 'gene_name', 'probe'))
		},
		top_qtl_probe = all_expr_vals %>%
			semi_join(probe_info %>% distinct(probe, n_var_per_probe) %>% filter(n_var_per_probe == 0), by = 'probe') %>%
			semi_join(best_point$no_snps %>% distinct(GN, probe, gene_name), by = c('GN', 'probe', 'gene_name'))
	)

	# same but using all probes (even ones with snps)
	# checked that this looks similar to so need to replot, either way it is correct to not use probes with snps for comparing expression data
	# we don't know how to scale expression with number of snps
	if (0) {
		unq_expr_vals_w_snp = list(
			avg_expr_per_gene = all_expr_vals_w_snp %>%
				group_by(GN, gene_name, strain) %>%
				summarise(expr_val = mean(expr_val), .groups = 'drop'),
			top_expr_probe = {
				max_expr_probes = all_expr_vals_w_snp %>%
					group_by(GN, gene_name, probe) %>%
					summarise(avg_expr = mean(expr_val), .groups = 'drop') %>%
					group_by(GN, gene_name) %>%
					slice_max(n = 1, order_by = avg_expr, with_ties = FALSE) %>% 
					ungroup 
				all_expr_vals_w_snp %>%
					semi_join(max_expr_probes, by = c('GN', 'gene_name', 'probe'))
			},
			top_qtl_probe = all_expr_vals_w_snp %>%
				semi_join(best_point$all_probes %>% distinct(GN, probe, gene_name), by = c('GN', 'probe', 'gene_name'))
		)
	}

	# check number of rows for each list
	# unq_expr_vals %>% map(nrow)
```


# Overall gene expression levels for all genes

```{r echo=TRUE, fig.width = 10, fig.height = 10, fig.align = 'center', cache = TRUE, eval = TRUE, warning = FALSE}
	to_plt = all_expr_vals

	p = to_plt %>%
		group_by(gene_name) %>%
		mutate(avg_expr_val = median(expr_val)) %>%
		ungroup %>%
		mutate(gene_name = fct_reorder(gene_name, desc(avg_expr_val))) %>%
		ggplot(aes(GN, expr_val, color = gene_name)) +
		geom_boxplot(outlier.shape = '.') + 
		geom_hline(data = ~.x %>% distinct(gene_name, avg_expr_val),
				   aes(yintercept = avg_expr_val)) + 
		geom_hline(yintercept = 8, linetype = 'dashed') + 
		scale_color_manual(values = gene_pal) + 
		scale_x_discrete(guide = guide_axis(check.overlap = TRUE)) + 
		facet_wrap(~gene_name, ncol = 7) +
		coord_cartesian(ylim = c(0, 20)) + 
		theme_half_open() + 
		theme(
			  # axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 8),
			  legend.position = 'none',
			  axis.text.x = element_blank(),
			  axis.ticks.x = element_blank()
		)
	p
	ggsave('test.pdf', p, w = 10, h = 10)

	# figure print
    ggsave('../pdfs/SuppFig14_Summaryofgeneexpressionlevels.pdf', p, w = 10, h = 10)
```

# Loci of interest

```{r echo=TRUE, fig.width = 6, fig.height = 4, fig.align = 'center', cache = TRUE, eval = TRUE}
	strain_info = readr::read_csv('../outs/strain_info.csv')
	# founder genotypes
  qtl_data = readRDS('../outs/qtl_data.rds')
	bxd_mark_gts = maxmarg(qtl_data$snp_probs[,'13'], minprob = 0.5) %>% 
		.[[1]] %>%
		as.data.frame %>%
		rownames_to_column(var = 'strain') %>%
		pivot_longer(cols = !strain, names_to = 'marker', values_to = 'fou_gt')
	# select marker at peak
	vcfs = list(snp = '../data/vep_annot/bxd_snp_indel.annot.vcf.gz', sv = '../data/vep_annot/bxd_svs.annot.vcf.gz')
	loci_of_int = list(
		qtl_peak = qtl_res %>% 
			filter(metric == '% expanded') %>%
			unite('marker', c('chr', 'pos', 'end')) %>%
			slice_max(LOD, n = 1) %>%
			select(locus = marker) %>%
			mutate(vcf = vcfs[['snp']]),
		closest_to_ci_center = bxd_mark_gts %>%
			distinct(marker) %>%
			separate('marker', c('chr', 'pos', 'end'), convert = TRUE, remove = FALSE) %>%
			mutate(dist_to_peak = abs(pos - ci_mid*1e6)) %>%
			slice_min(n = 1, order_by = dist_to_peak, with_ties = FALSE) %>%
			select(locus = marker) %>%
			mutate(vcf = vcfs[['snp']]),
		atg10_variant    = tibble(locus = 'chr13_91154245_91154245', vcf = vcfs[['snp']]),
		te_insertion     = tibble(locus = 'chr13_92348038_92348424', vcf = vcfs[['sv']]),
		msh3_splice_var1 = tibble(locus = 'chr13_92348451_92348451', vcf = vcfs[['snp']]),
		msh3_splice_var2 = tibble(locus = 'chr13_92348452_92348452', vcf = vcfs[['snp']])
		# this cryptic frame shift is most likely an artifact
		# cryptic_frame_shift = tibble(locus = 'chr13_92348043_92348043', vcf = vcfs[['snp']])
	) %>% map_df(~.x, .id = 'loc')
	loci_of_int
	# get genotypes at marker
	if (0) {
		# subset genotypes
		gts_of_int = bxd_mark_gts %>%
			mutate(fou_gt = recode(fou_gt, `1` = 'B', `2` = 'D'),
				   fou_gt = replace_na(fou_gt, 'miss')) %>%
			semi_join(loci_of_int, by = c('marker' = 'locus'))
		# check for markers
		gts_of_int %>% count(marker)
		# marker                      n
		# chr13_90420704_90420704   152
		# chr13_90440358_90440358   152
		# problem is that the atg10 variant is not included in the marker list using for QTL mapping
	} else {
		# debug
		# bcftools view ../data/vep_annot/bxd_svs.annot.vcf.gz chr13:92348030-92348434 | less -S
		
		# get raw genotypes from vcf
		# .x = 'chr13:92348038-92348424'; .y = '../data/vep_annot/bxd_svs.annot.vcf.gz'
		gts_of_int = loci_of_int %>% 
			separate('locus', c('chr', 'pos', 'end'), remove = FALSE) %>%
			rowwise %>%
			mutate(data = map2(sprintf('%s:%s-%s', chr, pos, end), vcf, function(.x, .y) {
				cmd = sprintf("bcftools query -f '[%%CHROM\t%%POS\t%%END\t%%REF\t%%ALT\t%%SAMPLE\t%%GT\t%%TGT\n]' %s -r %s", 
							  .y, 
							  .x)
				read_tsv(pipe(cmd), 
						 col_names = c('chr', 'pos', 'end', 'ref', 'alt', 'short_name', 'gt', 'tgt'),
						 col_types = cols(chr = 'c', pos = 'i', end = 'i', ref = 'c', alt = 'c', short_name = 'c', gt = 'c', tgt = 'c'))
			})) %>%
			ungroup
		gts_of_int = gts_of_int %>% select(loc, data) %>% unnest(data)
		
		# check genotypes
		# gts_of_int %>% pull(tgt) %>% unique
		# join real strain names
		gts_of_int = gts_of_int %>%
			left_join(strain_info %>% select(short_name, bxd_id), by = 'short_name') %>%
			mutate(bxd_id = if_else(loc == 'te_insertion', short_name, bxd_id)) %>%
			filter(!is.na(bxd_id)) %>%
			# count(loc)
			select(loc, chr, pos, end, ref, alt, strain = bxd_id, gt, tgt) %>%
			nest(gt_data = c(strain, gt, tgt))
	}
```


# Expression at peak LOD for top X eQTL genes + DNA repair genes

#. Only dataset/gene pairs were eQTL signal is significant
#. Only probes without snps

```{r echo=TRUE, fig.width = 14, fig.height = 10, fig.align = 'center', cache = TRUE, eval = TRUE}
	gene_gt_df = unq_expr_vals[['top_qtl_probe']] %>%
		left_join(gene_ord %>% select(gene_name, Rank), by = 'gene_name') %>%
		filter(Rank != other_lvl | gene_name %in% dna_repair_genes) %>%
		mutate(gene_name = fct_reorder(gene_name, as.integer(Rank))) %>%
		arrange(gene_name) %>%
		semi_join(eqtl_dsets_genes, by = c('GN', 'gene_name')) %>%
		# semi_join(qtl_dsets, by = 'GN') %>% # not necessary b/c eqtl_dsets_genes has this filtering already
		left_join(gts_of_int %>% filter(loc == 'qtl_peak') %>% unnest(gt_data) %>% select(strain, gt), by = 'strain')
	p_list = gene_gt_df %>%
		plt_avg_expr_per_fou_by_gn
	p = plot_grid(plotlist = p_list$p, nrow = 3)
	p

```

# DNA repair gene expression by founder and by GN

#. Only dataset/gene pairs were eQTL signal is significant

## Detailed view by dataset

```{r echo=TRUE, fig.width = 8, fig.height = 8, fig.align = 'center', cache = TRUE, eval = TRUE, warning = FALSE}
	# check
	# gene_gt_df %>% pull(strain) %>% unique
	p_list = gene_gt_df %>% 
		filter(gene_name %in% dna_repair_genes) %>% 
		plt_avg_expr_per_fou_by_gn

	# get rid of y-axis
	p_list = p_list %>%
		mutate(p = map(p, ~.x + theme(axis.title.y = element_blank())))

	# arrange
	# p = plot_grid(plotlist = p_list$p)
	p_top = (p_list %>% filter(panel_id == 'Ssbp2'))$p[[1]]
	p_bot = plot_grid(plotlist = (p_list %>% filter(panel_id != 'Ssbp2'))$p, nrow = 2)
	p = plot_grid(p_top, p_bot, ncol = 1, rel_heights = c(0.7, 1))
	p = plot_grid(ggdraw() + draw_text('Scaled gene expr.', x = 0.5, y = 0.5, size = 14, hjust = 0.5, vjust = 0.5, angle = 90),
				  p, nrow = 1, rel_widths = c(0.05, 1))
	p
	ggsave('test.pdf', p, w = 8, h = 8)

	# save figure
	ggsave('../pdfs/Suppl_Fig_19_Co-localization_b.pdf', p, w = 8, h = 8)

```

## Overall "B" vs "D" differences, Figure 4b

```{r echo=TRUE, fig.width = 6, fig.height = 4, fig.align = 'center', cache = TRUE, eval = TRUE, warning = FALSE}
	if (0) { # test for Msh3
		# dataset ordering by difference b/w B and D
		lvls = gene_gt_df %>% 
			filter(gene_name %in% dna_repair_genes) %>% 
			# filter(loc == 'qtl_peak') %>%
			filter(gt %in% c('0/0', '1/1')) %>%
			mutate(gt = recode(gt, `0/0` = 'B', `1/1` = 'D')) %>%
			group_by(gene_name, GN, gt) %>%
			summarise(expr_val = mean(expr_val), .groups = 'drop') %>%
			pivot_wider(id_cols = c(gene_name, GN), names_from = gt, values_from = expr_val) %>%
			mutate(diff = B - D) %>%
			arrange(desc(diff))
		p = gene_gt_df %>%
			filter(gene_name %in% dna_repair_genes) %>% 
			group_by(GN, gene_name) %>%
			mutate(expr_val = scale(expr_val, center = TRUE, scale = FALSE)[,1]) %>%
			ungroup %>%
			filter(gene_name == 'Msh3') %>%
			filter(gt %in% c('0/0', '1/1')) %>%
			mutate(gt = recode(gt, `0/0` = 'B', `1/1` = 'D')) %>%
			mutate(GN = fct_relevel(GN, lvls %>% filter(gene_name == 'Msh3') %>% pull(GN) %>% as.character)) %>%
			ggplot(aes(GN, expr_val, color = gt)) +
			geom_boxplot() + 
			scale_color_brewer(palette = 'Set1') + 
			coord_cartesian(ylim = c(-1.5, 1.5)) +
			theme_half_open() + 
			theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
				  axis.title.x = element_blank(),
				  legend.position = 'none',
				  plot.title = element_text(size = 10, hjust = 0.5))
	} else { # full plot
		# dataset ordering by difference b/w B and D
		lvls = gene_gt_df %>% 
			filter(gene_name %in% dna_repair_genes) %>% 
			filter(gt %in% c('0/0', '1/1')) %>%
			mutate(gt = recode(gt, `0/0` = 'B', `1/1` = 'D')) %>%
			group_by(gene_name, gt) %>%
			summarise(expr_val = mean(expr_val), .groups = 'drop') %>%
			pivot_wider(id_cols = c(gene_name), names_from = gt, values_from = expr_val) %>%
			mutate(diff = B - D) %>%
			arrange(desc(diff)) 

		# make plot
		p = gene_gt_df %>%
			filter(gene_name %in% dna_repair_genes) %>% 
			group_by(GN, gene_name) %>%
			mutate(expr_val = scale(expr_val, center = TRUE, scale = FALSE)[,1]) %>%
			ungroup %>%
			filter(gt %in% c('0/0', '1/1')) %>%
			mutate(gt = recode(gt, `0/0` = 'B', `1/1` = 'D')) %>%
			# mutate(gene_name = fct_relevel(gene_name, lvls %>% pull(gene_name) %>% as.character)) %>%
			mutate(gene_name = fct_relevel(gene_name, c('Atg10', 'Ccnh', 'Msh3', 'Ssbp2', 'Xrcc4'))) %>%
			ggplot(aes(gt, expr_val, color = gt)) +
			geom_boxplot() + 
			# stat_compare_means(method = 't.test', comparisons = list(c('B', 'D'))) + 
			# stat_compare_means(aes(label = ..p.signif..), method = 't.test') + 
			scale_color_brewer(palette = 'Set1') + 
			facet_wrap(~gene_name, nrow = 1, scales = 'free_x') + 
			theme_half_open() + 
			theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
				  axis.title.x = element_blank(),
				  legend.position = 'none',
				  plot.title = element_text(size = 10, hjust = 0.5)) +
			labs(y = 'Scaled gene expression')
	}
	p
	ggsave('test.pdf', p, w = 10, h = 8)

	# fig export
	p__bd_gene_expr = p
	ggsave('../pdfs/Figure4b.pdf', p, w = 10, h = 4)

```

